File: src/trades.c
#include "types.h"
#include <stdio.h>
#include <string.h>

/*
 * trades.c - Core trade execution primitives
 *
 * Pure mechanics: buy, sell, position tracking.
 * No strategy logic - just executes trades in memory.
 */

static const char *outcome_name(Outcome o) {
    switch (o) {
        case TEAM1: return "TEAM1";
        case DRAW:  return "DRAW";
        case TEAM2: return "TEAM2";
    }
    return "?";
}

void portfolio_init(Portfolio *p, double starting_cash) {
    memset(p, 0, sizeof(Portfolio));
    p->cash = starting_cash;
    printf("[PORTFOLIO] init: $%.2f\n", starting_cash);
}

static MatchPosition *get_or_create_match_pos(Portfolio *p, const char *event_id) {
    for (uint32_t i = 0; i < p->count; i++) {
        if (strcmp(p->positions[i].event_id, event_id) == 0) {
            return &p->positions[i];
        }
    }
    if (p->count >= MAX_MATCHES) {
        fprintf(stderr, "[ERR] portfolio full, max %d positions\n", MAX_MATCHES);
        return NULL;
    }
    MatchPosition *mp = &p->positions[p->count++];
    memset(mp, 0, sizeof(MatchPosition));
    snprintf(mp->event_id, sizeof(mp->event_id), "%s", event_id);
    return mp;
}

int buy(Portfolio *p, const Match *m, Outcome outcome, double dollars) {
    if (dollars <= 0) {
        fprintf(stderr, "[ERR] buy: amount must be positive (got $%.2f)\n", dollars);
        return -1;
    }
    if (dollars > p->cash) {
        fprintf(stderr, "[ERR] buy: need $%.2f, have $%.2f\n", dollars, p->cash);
        return -1;
    }
    double ask = m->markets[outcome].ask;
    if (ask <= 0) {
        fprintf(stderr, "[ERR] buy: no ask for %s %s\n", m->event_id, outcome_name(outcome));
        return -1;
    }

    double shares = dollars / ask;
    MatchPosition *mp = get_or_create_match_pos(p, m->event_id);
    if (!mp) return -1;

    mp->pos[outcome].shares += shares;
    mp->pos[outcome].cost_basis += dollars;
    p->cash -= dollars;

    printf("[BUY] %s %s: $%.2f @ %.3f -> %.2f shares\n",
           m->event_id, outcome_name(outcome), dollars, ask, shares);
    return 0;
}

int sell(Portfolio *p, const Match *m, Outcome outcome, double shares_to_sell) {
    if (shares_to_sell <= 0) {
        fprintf(stderr, "[ERR] sell: shares must be positive\n");
        return -1;
    }
    MatchPosition *mp = get_or_create_match_pos(p, m->event_id);
    if (!mp) return -1;

    Position *pos = &mp->pos[outcome];
    if (shares_to_sell > pos->shares) {
        fprintf(stderr, "[ERR] sell: have %.2f shares, tried %.2f\n", pos->shares, shares_to_sell);
        return -1;
    }
    double bid = m->markets[outcome].bid;
    if (bid <= 0) {
        fprintf(stderr, "[ERR] sell: no bid for %s %s\n", m->event_id, outcome_name(outcome));
        return -1;
    }

    double proceeds = shares_to_sell * bid;
    double fraction = shares_to_sell / pos->shares;
    double cost_sold = pos->cost_basis * fraction;

    pos->shares -= shares_to_sell;
    pos->cost_basis -= cost_sold;
    p->cash += proceeds;

    printf("[SELL] %s %s: %.2f shares @ %.3f -> $%.2f\n",
           m->event_id, outcome_name(outcome), shares_to_sell, bid, proceeds);
    return 0;
}

int sell_all(Portfolio *p, const Match *m, Outcome outcome) {
    MatchPosition *mp = get_or_create_match_pos(p, m->event_id);
    if (!mp || mp->pos[outcome].shares <= 0) {
        fprintf(stderr, "[ERR] sell_all: no position in %s %s\n", m->event_id, outcome_name(outcome));
        return -1;
    }
    return sell(p, m, outcome, mp->pos[outcome].shares);
}

Position *get_position(Portfolio *p, const char *event_id, Outcome outcome) {
    for (uint32_t i = 0; i < p->count; i++) {
        if (strcmp(p->positions[i].event_id, event_id) == 0) {
            return &p->positions[i].pos[outcome];
        }
    }
    return NULL;
}

double position_value(const Position *pos, double bid) {
    return pos->shares * bid;
}

double position_pnl(const Position *pos, double bid) {
    return position_value(pos, bid) - pos->cost_basis;
}

void portfolio_print(const Portfolio *p, const MatchBook *book) {
    printf("\n=== Portfolio ===\n");
    printf("Cash: $%.2f\n\n", p->cash);

    double total_value = p->cash;
    double total_cost = 0;

    for (uint32_t i = 0; i < p->count; i++) {
        const MatchPosition *mp = &p->positions[i];
        const Match *m = NULL;
        for (uint32_t j = 0; j < book->count; j++) {
            if (strcmp(book->matches[j].event_id, mp->event_id) == 0) {
                m = &book->matches[j];
                break;
            }
        }
        if (!m) {
            fprintf(stderr, "[WARN] event %s not in book\n", mp->event_id);
            continue;
        }

        int has_pos = 0;
        for (int o = 0; o < 3; o++) if (mp->pos[o].shares > 0) has_pos = 1;
        if (!has_pos) continue;

        printf("%s vs %s (%s)\n", m->team1, m->team2, m->date);
        const char *labels[] = {m->team1, "Draw", m->team2};
        for (int o = 0; o < 3; o++) {
            const Position *pos = &mp->pos[o];
            if (pos->shares > 0) {
                double val = position_value(pos, m->markets[o].bid);
                double pnl = position_pnl(pos, m->markets[o].bid);
                printf("  %-25s %.2f sh @ $%.2f | Val: $%.2f | PnL: %+.2f\n",
                       labels[o], pos->shares, pos->cost_basis, val, pnl);
                total_value += val;
                total_cost += pos->cost_basis;
            }
        }
        printf("\n");
    }

    printf("Position Value: $%.2f | Cost: $%.2f | PnL: %+.2f | Total: $%.2f\n",
           total_value - p->cash, total_cost, (total_value - p->cash) - total_cost, total_value);
}

File: src/trades.rs
use crate::types::{Match, MatchBook, MatchPosition, Outcome, Portfolio, Position, MAX_MATCHES};

pub fn portfolio_init(starting_cash: f64) -> Portfolio {
    println!("[PORTFOLIO] init: ${:.2}", starting_cash);
    Portfolio {
        positions: Vec::new(),
        cash: starting_cash,
    }
}

fn get_or_create_match_pos<'a>(p: &'a mut Portfolio, event_id: &str) -> Option<&'a mut MatchPosition> {
    // Check if position already exists
    let exists = p.positions.iter().any(|mp| mp.event_id == event_id);

    if exists {
        // Find and return the existing position
        p.positions.iter_mut().find(|mp| mp.event_id == event_id)
    } else {
        // Create new position
        if p.positions.len() >= MAX_MATCHES {
            eprintln!("[ERR] portfolio full, max {} positions", MAX_MATCHES);
            return None;
        }
        let mp = MatchPosition {
            event_id: event_id.to_string(),
            pos: Default::default(),
        };
        p.positions.push(mp);
        p.positions.last_mut()
    }
}

pub fn buy(p: &mut Portfolio, m: &Match, outcome: Outcome, dollars: f64) -> Result<(), String> {
    if dollars <= 0.0 {
        return Err(format!(
            "[ERR] buy: amount must be positive (got ${:.2})",
            dollars
        ));
    }
    if dollars > p.cash {
        return Err(format!(
            "[ERR] buy: need ${:.2}, have ${:.2}",
            dollars, p.cash
        ));
    }
    let ask = m.markets[outcome.as_index()].ask;
    if ask <= 0.0 {
        return Err(format!(
            "[ERR] buy: no ask for {} {}",
            m.event_id,
            outcome.name()
        ));
    }

    let shares = dollars / ask;
    let mp = get_or_create_match_pos(p, &m.event_id).ok_or("[ERR] buy: portfolio full")?;

    mp.pos[outcome.as_index()].shares += shares;
    mp.pos[outcome.as_index()].cost_basis += dollars;
    p.cash -= dollars;

    println!(
        "[BUY] {} {}: ${:.2} @ {:.3} -> {:.2} shares",
        m.event_id,
        outcome.name(),
        dollars,
        ask,
        shares
    );
    Ok(())
}

pub fn sell(
    p: &mut Portfolio,
    m: &Match,
    outcome: Outcome,
    shares_to_sell: f64,
) -> Result<(), String> {
    if shares_to_sell <= 0.0 {
        return Err("[ERR] sell: shares must be positive".to_string());
    }

    let mp = get_or_create_match_pos(p, &m.event_id).ok_or("[ERR] sell: portfolio full")?;

    let pos = &mp.pos[outcome.as_index()];
    if shares_to_sell > pos.shares {
        return Err(format!(
            "[ERR] sell: have {:.2} shares, tried {:.2}",
            pos.shares, shares_to_sell
        ));
    }
    let bid = m.markets[outcome.as_index()].bid;
    if bid <= 0.0 {
        return Err(format!(
            "[ERR] sell: no bid for {} {}",
            m.event_id,
            outcome.name()
        ));
    }

    let proceeds = shares_to_sell * bid;
    let fraction = shares_to_sell / pos.shares;
    let cost_sold = pos.cost_basis * fraction;

    let pos = &mut mp.pos[outcome.as_index()];
    pos.shares -= shares_to_sell;
    pos.cost_basis -= cost_sold;
    p.cash += proceeds;

    println!(
        "[SELL] {} {}: {:.2} shares @ {:.3} -> ${:.2}",
        m.event_id,
        outcome.name(),
        shares_to_sell,
        bid,
        proceeds
    );
    Ok(())
}

pub fn sell_all(p: &mut Portfolio, m: &Match, outcome: Outcome) -> Result<(), String> {
    let shares = {
        let mp = p.positions.iter().find(|mp| mp.event_id == m.event_id);
        match mp {
            Some(mp) if mp.pos[outcome.as_index()].shares > 0.0 => {
                mp.pos[outcome.as_index()].shares
            }
            _ => {
                return Err(format!(
                    "[ERR] sell_all: no position in {} {}",
                    m.event_id,
                    outcome.name()
                ));
            }
        }
    };
    sell(p, m, outcome, shares)
}

pub fn get_position<'a>(
    p: &'a Portfolio,
    event_id: &str,
    outcome: Outcome,
) -> Option<&'a Position> {
    p.positions
        .iter()
        .find(|mp| mp.event_id == event_id)
        .map(|mp| &mp.pos[outcome.as_index()])
}

pub fn position_value(pos: &Position, bid: f64) -> f64 {
    pos.shares * bid
}

pub fn position_pnl(pos: &Position, bid: f64) -> f64 {
    position_value(pos, bid) - pos.cost_basis
}

pub fn portfolio_print(p: &Portfolio, book: &MatchBook) {
    println!("\n=== Portfolio ===");
    println!("Cash: ${:.2}\n", p.cash);

    let mut total_value = p.cash;
    let mut total_cost = 0.0;

    for mp in &p.positions {
        let m = book.matches.iter().find(|m| m.event_id == mp.event_id);
        let m = match m {
            Some(m) => m,
            None => {
                eprintln!("[WARN] event {} not in book", mp.event_id);
                continue;
            }
        };

        let has_pos = mp.pos.iter().any(|p| p.shares > 0.0);
        if !has_pos {
            continue;
        }

        println!("{} vs {} ({})", m.team1, m.team2, m.date);
        let labels = [m.team1.as_str(), "Draw", m.team2.as_str()];
        for o in 0..3 {
            let pos = &mp.pos[o];
            if pos.shares > 0.0 {
                let val = position_value(pos, m.markets[o].bid);
                let pnl = position_pnl(pos, m.markets[o].bid);
                println!(
                    "  {:<25} {:.2} sh @ ${:.2} | Val: ${:.2} | PnL: {:+.2}",
                    labels[o], pos.shares, pos.cost_basis, val, pnl
                );
                total_value += val;
                total_cost += pos.cost_basis;
            }
        }
        println!();
    }

    println!(
        "Position Value: ${:.2} | Cost: ${:.2} | PnL: {:+.2} | Total: ${:.2}",
        total_value - p.cash,
        total_cost,
        (total_value - p.cash) - total_cost,
        total_value
    );
}
